woocommerce/woocommerce-ios
# Product inventory/SKU settings: barcode scanner for SKU

Part of #2407 

Sorry about the PR size going slightly over 500 diffs! I originally thought I'd PR the barcode scanner separately, but then it's hard to test the integration. About 100 diffs are in two xibs. When I started the barcode scanner back in June, I referred to some example projects in [Vision documentation](https://developer.apple.com/documentation/vision).

## Changes

- Created `BarcodeScannerViewController` that is meant to be embedded as a child view controller, for future reuse in inventory scanner (more details in https://github.com/woocommerce/woocommerce-ios/issues/2407). This starts up camera for live video stream, and makes Vision requests for barcode detection from video output frames at a throttled rate.
  - Added a `CGRect` helper to map the scan area (see screenshots below) to the video frame (which is often bigger than the device screen). This mapping took me quite some time, I tried adding comments but let me know if I can explain the calculation further. I tested this by adding a debug image view that shows the cropped image and made sure that it is the same as the scan area.
- Created `ProductSKUInputScannerViewController` that is just a container of `BarcodeScannerViewController` with a navigation bar title.
- Integration with product inventory settings:
  - In `ProductInventorySettingsActionHandler`, added `handleSKUFromBarcodeScanner` similar to the existing `handleSKUChange` to display SKU before validation (since the SKU comes from the barcode scanner and we need to update the UI manually).
  - In `ProductInventorySettingsViewController`, navigated to `ProductSKUInputScannerViewController` when the user taps on the scan CTA.

## Testing

Camera is only available on a device, please try on a device - iPad would be super helpful since I mostly tested on iPhone!

- Go to the products tab
- Tap on a product whose SKU is editable
- Tap on the inventory/SKU row
- Tap on the scan CTA in the accessory view of the SKU row --> should be prompted to use camera if you haven't used camera for other use cases in the app, tap yes in this case. No camera permission scenario will be handled separately as part of #2407.
- Scan any barcode that is available to you, make sure it's only scanned when the barcode is in the scan area --> after a barcode is detected, the barcode should be populated on the SKU row

## Example screencast/screenshots

Screencast: https://d.pr/free/v/C0DvGY

Light mode | Dark mode
-- | --
![IMG_7722](https://user-images.githubusercontent.com/1945542/98064568-94c39600-1e8d-11eb-8b07-b102ef68d2a7.PNG) | ![IMG_7723](https://user-images.githubusercontent.com/1945542/98064598-9d1bd100-1e8d-11eb-9419-df940d50afd9.PNG)

Landscape:
<img src="https://user-images.githubusercontent.com/1945542/98064605-a147ee80-1e8d-11eb-9354-578a9754ca39.PNG" width="500" />


Update release notes:
